File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        // int[] ans = new int[n];
        int[] prefixProduct = new int[n];
        // int[] suffixProduct = new int[n+1];
        // prefixProduct[0] = 1;
        // suffixProduct[n-1] = 1;
        

        int temp = 1;
        for(int i = 0; i<n; i++){
                prefixProduct[i] = temp;
                temp = prefixProduct[i]*nums[i];
        }
        temp = 1;

        for(int i = n-1; i>=0 ; i--){
                prefixProduct[i] = prefixProduct[i]*temp;
                temp = temp*nums[i];
        }

        // for(int i = 0; i<n; i++){
        //     ans[i] = prefixProduct[i]*suffixProduct[i];
        // }

        return prefixProduct;
    }
}  

File: Data Structures & Algorithms/top-k-elements-in-list/submission-1.kt
class Solution {
    fun topKFrequent(nums: IntArray, k: Int): IntArray {
        val ans = IntArray(k)
        val map = mutableMapOf<Int, Int>()

        // val pq = PriorityQueue<Pair<Int, Int>>(compareBy{it.second})
        
        //bucketSort
        val countArr = Array(nums.size+1){ArrayList<Int>()}
        

        for(num: Int in nums){
            map.put(num, map.getOrDefault(num, 0)+1)
        }

        map.forEach{(key, value) -> 
            // pq.add(Pair(key, value))
            // if(pq.size>k){
            //     pq.poll()
            // }

            // modified bucket sort
            val newList = countArr[value]
            newList.add(key)
            countArr[value] = newList
        }

        var count = k
        var idx = nums.size
        while(count>0 && idx > 0){
            if(countArr[idx] != null){
                for(num: Int in  countArr[idx]){
                    ans[count-1] = num
                    count--
                }
            }
            idx--
        }
        
        // for(i: Int in 0..k-1){
        //     ans[i] = pq.poll().first
        // }

        return ans
    }
}
